velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -1042,6 +1044,123 @@ TpchPlan TpchQueryBuilder::getQ19Plan() const {
   return context;
 }
 
+TpchPlan TpchQueryBuilder::getQ20Plan() const {
```

Comment:
Would you share the output of printPlanWithStats? I'd like to review the cardinalities of the plan nodes to understand why this is the best plan. It would be nice to include this information in the PR description.

Reply:
Adding to PR description, thanks

Execution time: 15.24s
Splits total: 320, finished: 320
-- OrderBy[s_name ASC NULLS LAST] -> s_name:VARCHAR, s_address:VARCHAR
   Output: 1804 rows (159.62KB, 1 batches), Cpu time: 7.72ms, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 8, Threads: 1
  -- HashJoin[LEFT SEMI s_suppkey=ps_suppkey] -> s_name:VARCHAR, s_address:VARCHAR
     Output: 1804 rows (0B, 12 batches), Cpu time: 38.34ms, Blocked wall time: 15.21s, Peak memory: 2.00MB, Memory allocations: 17
     HashBuild: Input: 58655 rows (1.46MB, 5315 batches), Output: 0 rows (0B, 0 batches), Cpu time: 38.20ms, Blocked wall time: 0ns, Peak memory: 2.00MB, Memory allocations: 17, Threads: 1
        distinctKey0       sum: 44483, count: 1, min: 44483, max: 44483
        queuedWallNanos    sum: 113.00us, count: 1, min: 113.00us, max: 113.00us
        rangeKey0          sum: 100000, count: 1, min: 100000, max: 100000
     HashProbe: Input: 1804 rows (39.37KB, 12 batches), Output: 1804 rows (0B, 12 batches), Cpu time: 141.00us, Blocked wall time: 15.21s, Peak memory: 0B, Memory allocations: 0, Threads: 1
        dataSourceLazyWallNanos          sum: 11.48ms, count: 48, min: 9.00us, max: 532.00us
        dynamicFiltersProduced           sum: 1, count: 1, min: 1, max: 1
        queuedWallNanos                  sum: 92.00us, count: 1, min: 92.00us, max: 92.00us
        replacedWithDynamicFilterRows    sum: 1804, count: 12, min: 62, max: 193
    -- HashJoin[INNER s_nationkey=n_nationkey] -> s_name:VARCHAR, s_address:VARCHAR, s_suppkey:BIGINT
       Output: 1804 rows (39.37KB, 12 batches), Cpu time: 291.00us, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 2
       HashBuild: Input: 1 rows (672B, 1 batches), Output: 0 rows (0B, 0 batches), Cpu time: 153.00us, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 2, Threads: 1
          distinctKey0       sum: 2, count: 1, min: 2, max: 2
          queuedWallNanos    sum: 141.00us, count: 1, min: 141.00us, max: 141.00us
          rangeKey0          sum: 2, count: 1, min: 2, max: 2
       HashProbe: Input: 1804 rows (1014.25KB, 12 batches), Output: 1804 rows (39.37KB, 12 batches), Cpu time: 138.00us, Blocked wall time: 0ns, Peak memory: 0B, Memory allocations: 0, Threads: 1
          dynamicFiltersProduced           sum: 1, count: 1, min: 1, max: 1
          replacedWithDynamicFilterRows    sum: 1804, count: 12, min: 62, max: 193
      -- TableScan[table: supplier] -> s_nationkey:BIGINT, s_suppkey:BIGINT, s_name:VARCHAR, s_address:VARCHAR
         Input: 1804 rows (1014.25KB, 0 batches), Raw Input: 100000 rows (47.46MB), Output: 1804 rows (1014.25KB, 12 batches), Cpu time: 14.23ms, Blocked wall time: 0ns, Peak memory: 3.00MB, Memory allocations: 307, Threads: 1, Splits: 70
            dataSourceWallNanos       sum: 5.28ms, count: 82, min: 3.00us, max: 1.61ms
            dynamicFiltersAccepted    sum: 2, count: 2, min: 1, max: 1
            localReadBytes            sum: 0B, count: 1, min: 0B, max: 0B
            numLocalRead              sum: 0, count: 1, min: 0, max: 0
            numPrefetch               sum: 0, count: 1, min: 0, max: 0
            numRamRead                sum: 0, count: 1, min: 0, max: 0
            numStorageRead            sum: 0, count: 1, min: 0, max: 0
            prefetchBytes             sum: 0B, count: 1, min: 0B, max: 0B
            ramReadBytes              sum: 0B, count: 1, min: 0B, max: 0B
            skippedSplitBytes         sum: 0B, count: 1, min: 0B, max: 0B
            skippedSplits             sum: 0, count: 1, min: 0, max: 0
            skippedStrides            sum: 0, count: 1, min: 0, max: 0
            storageReadBytes          sum: 0B, count: 1, min: 0B, max: 0B
      -- TableScan[table: nation, range filters: [(name, Filter(BytesValues, deterministic, null not allowed))]] -> n_nationkey:BIGINT, n_name:VARCHAR
         Input: 1 rows (672B, 0 batches), Raw Input: 25 rows (15.33KB), Output: 1 rows (672B, 1 batches), Cpu time: 906.00us, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 29, Threads: 1, Splits: 10
            dataSourceLazyWallNanos    sum: 71.00us, count: 2, min: 21.00us, max: 50.00us
            dataSourceWallNanos        sum: 287.00us, count: 11, min: 2.00us, max: 231.00us
            localReadBytes             sum: 0B, count: 1, min: 0B, max: 0B
            numLocalRead               sum: 0, count: 1, min: 0, max: 0
            numPrefetch                sum: 0, count: 1, min: 0, max: 0
            numRamRead                 sum: 0, count: 1, min: 0, max: 0
            numStorageRead             sum: 0, count: 1, min: 0, max: 0
            prefetchBytes              sum: 0B, count: 1, min: 0B, max: 0B
            ramReadBytes               sum: 0B, count: 1, min: 0B, max: 0B
            skippedSplitBytes          sum: 0B, count: 1, min: 0B, max: 0B
            skippedSplits              sum: 0, count: 1, min: 0, max: 0
            skippedStrides             sum: 0, count: 1, min: 0, max: 0
            storageReadBytes           sum: 0B, count: 1, min: 0B, max: 0B
    -- HashJoin[INNER l_partkey=ps_partkey AND l_suppkey=ps_suppkey, filter: gt(cast ROW["ps_availqty"] as DOUBLE,ROW["filter_qty"])] -> ps_suppkey:BIGINT
       Output: 58655 rows (1.46MB, 5315 batches), Cpu time: 386.71ms, Blocked wall time: 2.91s, Peak memory: 6.00MB, Memory allocations: 10698
       HashBuild: Input: 86204 rows (74.94MB, 803 batches), Output: 0 rows (0B, 0 batches), Cpu time: 163.05ms, Blocked wall time: 0ns, Peak memory: 5.00MB, Memory allocations: 50, Threads: 1
          distinctKey0       sum: 21552, count: 1, min: 21552, max: 21552
          distinctKey1       sum: 57814, count: 1, min: 57814, max: 57814
          queuedWallNanos    sum: 87.00us, count: 1, min: 87.00us, max: 87.00us
          rangeKey0          sum: 1999991, count: 1, min: 1999991, max: 1999991
          rangeKey1          sum: 100000, count: 1, min: 100000, max: 100000
       HashProbe: Input: 5441669 rows (187.01MB, 5315 batches), Output: 58655 rows (1.46MB, 5315 batches), Cpu time: 223.66ms, Blocked wall time: 2.91s, Peak memory: 1.00MB, Memory allocations: 10648, Threads: 1
          queuedWallNanos    sum: 45.00us, count: 1, min: 45.00us, max: 45.00us
      -- Project[expressions: (l_partkey:BIGINT, ROW["l_partkey"]), (l_suppkey:BIGINT, ROW["l_suppkey"]), (filter_qty:DOUBLE, multiply(0.5,ROW["sum_qty"]))] -> l_partkey:BIGINT, l_suppkey:BIGINT, filter_qty:DOUBLE
         Output: 5441669 rows (187.01MB, 5315 batches), Cpu time: 57.83ms, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 5315, Threads: 1
        -- Aggregation[FINAL [l_partkey, l_suppkey] sum_qty := sum("sum_qty")] -> l_partkey:BIGINT, l_suppkey:BIGINT, sum_qty:DOUBLE
           Output: 5441669 rows (187.02MB, 5315 batches), Cpu time: 2.11s, Blocked wall time: 0ns, Peak memory: 248.00MB, Memory allocations: 13388, Threads: 1
          -- LocalPartition[REPARTITION] -> l_partkey:BIGINT, l_suppkey:BIGINT, sum_qty:DOUBLE
             Output: 17783894 rows (611.89MB, 8695 batches), Cpu time: 89.04ms, Blocked wall time: 10.12s, Peak memory: 0B, Memory allocations: 0
             LocalPartition: Input: 8891947 rows (305.95MB, 8695 batches), Output: 8891947 rows (305.95MB, 0 batches), Cpu time: 52.44ms, Blocked wall time: 631.91ms, Peak memory: 0B, Memory allocations: 0, Threads: 1
                queuedWallNanos    sum: 4.63ms, count: 159, min: 17.00us, max: 315.00us
             LocalExchange: Input: 8891947 rows (305.95MB, 0 batches), Output: 8891947 rows (305.95MB, 8695 batches), Cpu time: 36.60ms, Blocked wall time: 9.49s, Peak memory: 0B, Memory allocations: 0, Threads: 1
                queuedWallNanos    sum: 1.30ms, count: 22, min: 30.00us, max: 69.00us
            -- Aggregation[PARTIAL [l_partkey, l_suppkey] sum_qty := sum(ROW["l_quantity"])] -> l_partkey:BIGINT, l_suppkey:BIGINT, sum_qty:DOUBLE
               Output: 8891947 rows (305.95MB, 8695 batches), Cpu time: 6.74s, Blocked wall time: 0ns, Peak memory: 52.00MB, Memory allocations: 47982, Threads: 1
                  flushRowCount            sum: 8781673, count: 24, min: 365315, max: 366668
                  partialAggregationPct    sum: 2328, count: 24, min: 97, max: 97
              -- TableScan[table: lineitem, range filters: [(shipdate, BytesRange: [1994-01-01, 1994-12-31] no nulls)]] -> l_shipdate:VARCHAR, l_suppkey:BIGINT, l_partkey:BIGINT, l_quantity:DOUBLE
                 Input: 9099165 rows (749.18MB, 0 batches), Raw Input: 59986052 rows (574.16MB), Output: 9099165 rows (280.76MB, 6011 batches), Cpu time: 6.68s, Blocked wall time: 0ns, Peak memory: 20.00MB, Memory allocations: 18948, Threads: 1, Splits: 80
                    dataSourceLazyWallNanos    sum: 10.35s, count: 30055, min: 4.00us, max: 7.30ms
                    dataSourceWallNanos        sum: 6.86s, count: 6091, min: 20.00us, max: 48.01ms
                    loadedToValueHook          sum: 9099165, count: 6011, min: 489, max: 1827
                    localReadBytes             sum: 0B, count: 1, min: 0B, max: 0B
                    numLocalRead               sum: 0, count: 1, min: 0, max: 0
                    numPrefetch                sum: 0, count: 1, min: 0, max: 0
                    numRamRead                 sum: 0, count: 1, min: 0, max: 0
                    numStorageRead             sum: 0, count: 1, min: 0, max: 0
                    prefetchBytes              sum: 0B, count: 1, min: 0B, max: 0B
                    ramReadBytes               sum: 0B, count: 1, min: 0B, max: 0B
                    skippedSplitBytes          sum: 0B, count: 1, min: 0B, max: 0B
                    skippedSplits              sum: 0, count: 1, min: 0, max: 0
                    skippedStrides             sum: 0, count: 1, min: 0, max: 0
                    storageReadBytes           sum: 0B, count: 1, min: 0B, max: 0B
      -- HashJoin[LEFT SEMI ps_partkey=p_partkey] -> ps_partkey:BIGINT, ps_suppkey:BIGINT, ps_availqty:INTEGER
         Output: 86204 rows (74.94MB, 803 batches), Cpu time: 68.93ms, Blocked wall time: 952.35ms, Peak memory: 20.00MB, Memory allocations: 27
         HashBuild: Input: 21551 rows (1.12MB, 204 batches), Output: 0 rows (0B, 0 batches), Cpu time: 57.99ms, Blocked wall time: 0ns, Peak memory: 20.00MB, Memory allocations: 27, Threads: 1
            distinctKey0       sum: 21552, count: 1, min: 21552, max: 21552
            queuedWallNanos    sum: 111.00us, count: 1, min: 111.00us, max: 111.00us
            rangeKey0          sum: 1999991, count: 1, min: 1999991, max: 1999991
         HashProbe: Input: 86204 rows (74.94MB, 803 batches), Output: 86204 rows (74.94MB, 803 batches), Cpu time: 10.94ms, Blocked wall time: 952.35ms, Peak memory: 0B, Memory allocations: 0, Threads: 1
            dataSourceLazyWallNanos          sum: 254.50ms, count: 3212, min: 2.00us, max: 1.37ms
            dynamicFiltersProduced           sum: 1, count: 1, min: 1, max: 1
            queuedWallNanos                  sum: 43.00us, count: 1, min: 43.00us, max: 43.00us
            replacedWithDynamicFilterRows    sum: 86204, count: 803, min: 8, max: 196
        -- TableScan[table: partsupp] -> ps_partkey:BIGINT, ps_suppkey:BIGINT, ps_availqty:INTEGER
           Input: 86204 rows (74.94MB, 0 batches), Raw Input: 8000000 rows (123.26MB), Output: 86204 rows (74.94MB, 803 batches), Cpu time: 1.70s, Blocked wall time: 0ns, Peak memory: 9.00MB, Memory allocations: 2681, Threads: 1, Splits: 80
              dataSourceWallNanos       sum: 1.73s, count: 883, min: 3.00us, max: 6.89ms
              dynamicFiltersAccepted    sum: 1, count: 1, min: 1, max: 1
              localReadBytes            sum: 0B, count: 1, min: 0B, max: 0B
              numLocalRead              sum: 0, count: 1, min: 0, max: 0
              numPrefetch               sum: 0, count: 1, min: 0, max: 0
              numRamRead                sum: 0, count: 1, min: 0, max: 0
              numStorageRead            sum: 0, count: 1, min: 0, max: 0
              prefetchBytes             sum: 0B, count: 1, min: 0B, max: 0B
              ramReadBytes              sum: 0B, count: 1, min: 0B, max: 0B
              skippedSplitBytes         sum: 0B, count: 1, min: 0B, max: 0B
              skippedSplits             sum: 0, count: 1, min: 0, max: 0
              skippedStrides            sum: 0, count: 1, min: 0, max: 0
              storageReadBytes          sum: 0B, count: 1, min: 0B, max: 0B
        -- TableScan[table: part, remaining filter: (like(ROW["name"],"forest%"))] -> p_partkey:BIGINT, p_name:VARCHAR
           Input: 21551 rows (122.78MB, 0 batches), Raw Input: 2000000 rows (406.29MB), Output: 21551 rows (1.12MB, 204 batches), Cpu time: 853.61ms, Blocked wall time: 0ns, Peak memory: 8.00MB, Memory allocations: 3481, Threads: 1, Splits: 80
              dataSourceLazyWallNanos    sum: 230.17ms, count: 408, min: 8.00us, max: 5.07ms
              dataSourceWallNanos        sum: 755.63ms, count: 284, min: 6.00us, max: 6.12ms
              localReadBytes             sum: 0B, count: 1, min: 0B, max: 0B
              numLocalRead               sum: 0, count: 1, min: 0, max: 0
              numPrefetch                sum: 0, count: 1, min: 0, max: 0
              numRamRead                 sum: 0, count: 1, min: 0, max: 0
              numStorageRead             sum: 0, count: 1, min: 0, max: 0
              prefetchBytes              sum: 0B, count: 1, min: 0B, max: 0B
              ramReadBytes               sum: 0B, count: 1, min: 0B, max: 0B
              skippedSplitBytes          sum: 0B, count: 1, min: 0B, max: 0B
              skippedSplits              sum: 0, count: 1, min: 0, max: 0
              skippedStrides             sum: 0, count: 1, min: 0, max: 0
              storageReadBytes           sum: 0B, count: 1, min: 0B, max: 0B